Using-GTSAM-EXPORT.md

```diff
@@ -0,0 +1,41 @@
+Unfortunately, using `GTSAM_EXPORT` incorrectly often does not cause a compiler or linker error in the library that is being compiled, but only when you try to use that DLL in a different library.  For example, an error in gtsam/base will often show up when compiling the check_base_program or the MATLAB wrapper, but not when compiling/linking gtsam itself.  The most common errors will say something like:
+
+```
+Error	LNK2019	unresolved external symbol "public: void __cdecl gtsam::SO3::print(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (?print@SO3@gtsam@@QEBAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "public: static void __cdecl gtsam::Testable<class gtsam::SO3>::Print(class gtsam::SO3 const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?Print@?$Testable@VSO3@gtsam@@@gtsam@@SAXAEBVSO3@2@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)	check_geometry_program	C:\AFIT\lib\gtsam\build\gtsam\geometry\tests\testSO3.obj
+```
+
+Sorry for the long error message there, but I want to quickly decode it to help understand what is going on.  First, there is an unresolved symbol `gtsam::SO3::print`.  This can occur because _either_ `GTSAM_EXPORT` was not added to the print function definition when it should have been, _OR_ because `GTSAM_EXPORT` was added to the print function definition when it is fully declared in the header.  This error was detected while compiling `check_geometry_program` and pulling in `...\testSO3.obj`.  Specifically, within the function call `gtsam::Testabl<class gtsam::SO3>::Print (...)`.  Note that this error did _not_ occur when compiling the library that actually has SO3 inside of it.
+
+## But Why?
+To me, the behavior of GTSAM_EXPORT was very baffling.  However, I believe it can be explained by understanding the rules that MSVC operates under.  
+
+First, we need to understand exactly what `GTSAM_EXPORT` does.  When you use `GTSAM_EXPORT` on a function (or class) definition, it actually does two things.  First, when you are building the library, it turns into a "dllexport" command, telling the compiler that this function should go into the DLL and be visible externally.  When you pull in that same header for a different library, it switches to a "dllimport" command, telling the compiler it should find this function in a DLL somewhere.  This leads to the first rule the compiler uses.  (Note that I say "compiler rules" when the rules may actually be in the linker, but I am conflating the two terms here when I speak of the "compiler rules".) 
+
+***Compiler Rule #1*** If a `dllimport` command is used in defining a function or class, that function or class _must_ be found in a DLL.
+
+Rule #1 doesn't seem very bad, until you combine it with rule #2
+
+***Compiler Rule #2*** Anything defined in a header file is not included in a DLL.
```

Comment:
> Anything defined in a header file is not included in a DLL.

I'm not 100% sure of this... it is NOT exported if you mark it explicitly with `inline`. Otherwise, I'm not sure if there is a well-defined, standardized expected behavior (?). 

I'll kindly request the knowledge of @jolting here just in case he can help us to be accurate in the description...
@jolting

Reply:
Please note that my expertise on DLLimport/export is what I have learned playing with GTSAM for the last week or so getting it to compile and looking up things online.  So, I am theorizing to the best of my knowledge, but I certainly am not an expert.

That said, I have seen several website mention that if you declare a function in a header inside a class definition, the compiler automatically makes it an "inline" function.  If it was not inline and you included this class definition in more than one file, then you would get a "multiple definitions of the function 'doit'" error.  So this one is essentially marked inline and will not be put in the DLL.

Once again, I should probably change "define" to declare.  My im-precision may be causing confusion. :(

I guess it may be possible that a non-inline function is declared in a header file and included in the DLL, but that seems like a corner case (due to the "multiple instances" compilation problem inherent with non-inline functions) so I don't think we need to worry about that too much.